_posts/2016-08-03-testing-spark-streaming-integration-testing-with-docker-compose.md

```diff
@@ -0,0 +1,244 @@
+
+### Controlling the lifecycle
+
+Managing how and when a process should start, stop or move into different states is part of the process lifecycle management. Let's see some considerations about this management when integration testing.
+
+- If the process is expensive to spin up, from time or space point of view, keeping the process started for the whole test suite could be convenient.
+- That might be problematic if the process is stateful. If that's the case we need to be sure that the data is partitioned between tests, so tests won't step on each other toes.
+- If isolating data becomes too complicated, we can dispose the process associated with every test or logical group of tests, in order to be sure that we work with a clean slate.
+- Another approach is deleting the data generated by the test. Deleting the data after the test has the following problems: harder diagnostic if the test fails and risk of not deleting data if the test fails catastrophically in the middle. My preferred approach is deleting the data before starting every test.
+- Being sure that a process has 'spun up' completely is hard to say. Some processes could be up with a PID assigned but not ready to receive messages until some warm up happens. That complicates massively the management as there is no general solution.
+
+### Coupling testing dependencies with your build system.
+
+In this post we're going to explore Docker Compose to control external dependencies. Docker Compose is a lightweight way of packaging applications, but even that it takes some time to start the containers (often related with the startup time of the processes themselves).
+
+Therefore we'll go with the approach of starting the containers once per test suite.
+
+plugins.sbt
+```scala
+addSbtPlugin("com.tapad" % "sbt-docker-compose" % "1.0.11")
+```
+
+build.sbt
+```scala
+lazy val dockerComposeTag = "DockerComposeTag"
+
+enablePlugins(DockerComposePlugin)
+
+testOptions in Test += Tests.Argument(TestFrameworks.ScalaTest, "-l", dockerComposeTag),
+  composeFile := baseDirectory.value + "/docker/sbt-docker-compose.yml",
+  testTagsToExecute := dockerComposeTag
+  )
+```
+
+I've copied the most relevant bits from our example. As you can see, we're using [sbt-docker-compose](https://github.com/Tapad/sbt-docker-compose) library. That means that we're coupling our tests (at least their dependencies) with the build system ([sbt](http://www.scala-sbt.org/)). That could be a problem as we're locked-in our solution to this particular build provider, but as usual, there is a trade-off in every technical decision.
+
+Every test that is tagged with `DockerComposeTag` will be executed when running `sbt dockerComposeTest`. This command will spin up and tear down the containers defined in `sbt-docker-compose.yml`:
```

Comment:
although spin up and tear down is correct grammatically - you want to say "set up and tear down" instead  - tests related terminologies